### ISSUE TITLE
TLS Secrets in Ingresses aren't loaded when the Secret is created after the Ingress 

### ISSUE THREAD

**Issue by maelvls** [cilium/cilium-service-mesh-beta#22]
I was writing the example "TLS with cert-manager" (#20) and realized that the Cilium Service Mesh with cilium v1.11.0-beta.1 does not seem to be loading the TLS Secret when the TLS Secret is not immediately present (it takes a few seconds for cert-manager to create the Secret after the Ingress is created).

**Reproducing the issue:**

To reproduce, you will have to install Cilium Service Mesh on a Kind cluster, and then install cert-manager. Here are the commands to create the Kind cluster (using kind v0.11.1) and install Cilium v1.11.0-beta.1:

```sh

kind create cluster --config /dev/stdin <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  disableDefaultCNI: true
EOF

cilium install --version -service-mesh:v1.11.0-beta.1 --config enable-envoy-config=true --kube-proxy-replacement=probe
cilium hubble enable

# https://kind.sigs.k8s.io/docs/user/loadbalancer/
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/master/manifests/namespace.yaml
kubectl create secret generic -n metallb-system memberlist --from-literal=secretkey="$(openssl rand -base64 128)"
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/master/manifests/metallb.yaml
kubectl wait -n metallb-system --for=condition=Ready pods --all
NET=$(docker network inspect kind | jq '.[].IPAM.Config[0] | select(.Gateway) | .Subnet' -r | sed 's|/16||') && \
  kubectl -n metallb-system create configmap config --dry-run=client -oyaml --from-file=config=/dev/stdin \
    <<< "address-pools: [{name: default, protocol: layer2, addresses: ['${NET/%.0/.200}-${NET/%.0/.250}']}]" | kubectl apply -f-
```

Then, let us install cert-manager with an issuer `ca-issuer`:

```sh
helm repo add jetstack https://charts.jetstack.io
helm upgrade --install cert-manager jetstack/cert-manager --version v1.7.1 --namespace cert-manager --set installCRDs=true --create-namespace
kubectl apply -f- <<EOF
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: self-signed
  namespace: default
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: ca
  namespace: default
spec:
  isCA: true
  privateKey:
    algorithm: ECDSA
    size: 256
  secretName: ca
  commonName: ca
  issuerRef:
    name: self-signed
    kind: Issuer
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: ca-issuer
  namespace: default
spec:
  ca:
    secretName: ca
EOF
```

Finally, let us install a demo application, and configure your DNS to resolve the domains ookinfo.cilium.rocks and hipstershop.cilium.rocks:

```
kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.11/samples/bookinfo/platform/kube/bookinfo.yaml
kubectl apply -f https://raw.githubusercontent.com/cilium/cilium-service-mesh-beta/e5c1f5b/kubernetes-ingress/tls-ingress.yaml
sudo perl -ni -e 'print if !/\.cilium\.rocks$/d' /etc/hosts; sudo tee -a /etc/hosts \
  <<<"$(kubectl get svc/cilium-ingress-tls-ingress -o=jsonpath='{.status.loadBalancer.ingress[0].ip}') bookinfo.cilium.rocks hipstershop.cilium.rocks"
kubectl wait --for=condition=Ready pods -n default --all
```

At this point, `demo-cert` does not exist yet since we haven't annotated the Ingress resource with `cert-manager.io/issuer=ca-issuer`. Let's check that the Secret is missing:

```console
$ kubectl get secret demo-cert
Error from server (NotFound): secrets "demo-cert" not found
```

Weridly enough, Cilium Service Mesh seems to have configured the Ingress to listen on 443 over HTTP (instead of HTTPS), which means `curl` complains about not receiving the Hello Server (instead, it receives `HTTP/1.1 200 OK`):

```console
# Over TLS:
$ curl -sS -o /dev/null -D /dev/stdout https://bookinfo.cilium.rocks:443/details/1
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number

# Over HTTP:
$ curl -sS -o /dev/null -D /dev/stdout https://bookinfo.cilium.rocks:443/details/1
HTTP/1.1 200 OK
content-type: application/json
server: envoy
date: Fri, 18 Feb 2022 14:59:56 GMT
content-length: 178
x-envoy-upstream-service-time: 2
```

Now, let us create the Secret `demo-cert` by asking cert-manager:

```sh
kubectl annotate ingress tls-ingress cert-manager.io/issuer=ca-issuer
```

We can check that the certificate has been properly created:

```
$ kubectl get secret demo-cert
NAME                               TYPE                                  DATA   AGE
demo-cert                          kubernetes.io/tls                     3      36s
```

I waited for a bit, expecting Cilium Service Mesh to "load" `demo-cert`, but it doesn't seem to be doing so and it keeps serving over HTTP on port 443.

I noticed that you need to recreate or update the Ingress resource to get Cilium Service Mesh to take `demo-cert` into account:

```sh
kubectl delete -f https://raw.githubusercontent.com/cilium/cilium-service-mesh-beta/e5c1f5b/kubernetes-ingress/tls-ingress.yaml
kubectl apply -f https://raw.githubusercontent.com/cilium/cilium-service-mesh-beta/e5c1f5b/kubernetes-ingress/tls-ingress.yaml
```

Two questions:

- Is it possible to enable a "strict" HTTPS-only mode on the Ingress? I would never expect 443 to be serving HTTP.
- Is it expected that the TLS Secret `demo-cert` doesn't get dynamically loaded by Cilium Service Mesh when it comes up?

**Versions:**

- Cilium  v1.11.0-beta.1
- Kernel 5.13.0-28-generic
- Kubernetes v1.22.4 (kind v0.11.1)
-  Sysdump: [cilium-sysdump-20220218-160816.zip](https://github.com/cilium/cilium-service-mesh-beta/files/8098201/cilium-sysdump-20220218-160816.zip)

I agree to follow this project's Code of Conduct.

**Comment by sayboras** [comment]
> Is it possible to enable a "strict" HTTPS-only mode on the Ingress? I would never expect 443 to be serving HTTP.
This seems to be a bug from our side. Currently, if the secret is not available, we just silently ignore the error, and didn't configure TLS downstream accordingly i.e. listener is with http only. 

> Is it expected that the TLS Secret demo-cert doesn't get dynamically loaded by Cilium Service Mesh when it comes up?
Reconciliation is based on CRUD events right now, there is race condition between cert manager and cilium. For the above case, you can make a dummy modification after annotation is added.

Currently, there is discussion to move secret lookup to runtime, which can help to solve both problems. Thanks a lot for your detailed steps 💯 .

**Comment by sayboras** [comment]
This should be fixed as part of v1.12.0-rc2, thanks.
